pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,393 @@
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *pluginProvider) IsPlugin() bool {
+	return true
+}
+
+// isImageAllowed returns true if the image matches against the list of allowed matches by the plugin.
+func (p *pluginProvider) isImageAllowed(image string) bool {
+	for _, matchImage := range p.matchImages {
+		if matched, _ := credentialprovider.URLsMatchStr(matchImage, image); matched {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getCachedCredentials returns a credentialprovider.DockerConfig if cached from the plugin.
+func (p *pluginProvider) getCachedCredentials(image string) (credentialprovider.DockerConfig, bool, error) {
+	obj, found, err := p.cache.GetByKey(globalCacheKey)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	registry, err := parseRegistry(image)
+	if err != nil {
+		return nil, false, err
+	}
+
+	obj, found, err = p.cache.GetByKey(registry)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	obj, found, err = p.cache.GetByKey(image)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	return nil, false, nil
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a provided CRI auth plugin based
+// on it's name provided in CRIAuthPluginConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	args         []string
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary in the following format:
+//
+//  $ <plugin-name> get-credentials arg1=value1 arg2=value2
+//
+// The plugin is expected to receive the CRIAuthPluginRequest API via stdin and is expected
+// to return the CRIAuthPluginResponse via stdout.
+func (e *execPlugin) ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error) {
+	pluginBin := filepath.Join(e.pluginBinDir, e.name)
+	args := []string{"get-credentials"}
+	args = append(args, e.args...)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
+
+	// Use a catch-all timeout of 1 minute for all exec-based plugins, this should leave enough
+	// head room in case a plugin needs to retry a failed request while ensuring an exec plugin
+	// does not run forever. In the future we may want this timeout to be tweakable from the plugin
+	// config file.
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```

Comment:
plumb a context arg into ExecPlugin here and wrap it to add a 1 minute backstop timeout

```
ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
```